gemini/use-cases/retrieval-augmented-generation/scalable_rag_with_bigframes.ipynb

```diff
@@ -0,0 +1,1099 @@
+        "    embedding_field=\"embedding_result\",\n",
+        "    content_field=\"text_chunk\",\n",
+        ")"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "U0fvQQwKrvoj"
+      },
+      "outputs": [],
+      "source": [
+        "# Perform similarity search and look at the most relevant documents\n",
+        "search_query = \"how do I read a csv file with python?\"  # @param {type:\"string\"}\n",
+        "results = bq_store.similarity_search(search_query)\n",
+        "text_results = [x.page_content for x in results]\n",
+        "text_results"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "### Answer Generation\n",
+        "\n",
+        "Now we can put everything together and use an LLM to answer a question based on Stack Overflow data!\n",
+        "\n",
+        "We are going to use LangChain and the [`RetrievalQA` chain](https://python.langchain.com/docs/versions/migrating_chains/retrieval_qa/) to build a very simple RAG chain.\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {},
+      "outputs": [],
+      "source": [
+        "from langchain import hub\n",
+        "from langchain.chains import create_retrieval_chain\n",
+        "from langchain.chains.combine_documents import create_stuff_documents_chain\n",
+        "from langchain_google_vertexai import ChatVertexAI\n",
+        "\n",
+        "# Convert the BigQuery VectorStore to a LangChain retriever\n",
+        "langchain_retriever = bq_store.as_retriever()\n",
+        "\n",
+        "# Init the VertexAI LLM\n",
+        "llm = ChatVertexAI(model_name=\"gemini-1.5-flash\")\n",
+        "\n",
+        "# See full prompt at https://smith.langchain.com/hub/langchain-ai/retrieval-qa-chat\n",
+        "retrieval_qa_chat_prompt = hub.pull(\"langchain-ai/retrieval-qa-chat\")\n",
+        "\n",
+        "combine_docs_chain = create_stuff_documents_chain(llm, retrieval_qa_chat_prompt)\n",
+        "rag_chain = create_retrieval_chain(langchain_retriever, combine_docs_chain)\n",
+        "\n",
+        "print(rag_chain.invoke({\"input\": search_query})[\"answer\"])"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "L2o86ME-rvoj"
+      },
+      "source": [
+        "## Scaling data processing to Terabytes: BigFrame Remote Functions"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "unwXv4-Urvoj"
+      },
+      "source": [
+        "Sometimes data is too large to run local process when running custom Python functions.\n",
+        "In fact, every time we convert a series or a DataFrame to pandas using `to_pandas()`, the data is loaded into memory.\n",
+        "\n",
+        "To be able to run large datasets processes remotely, we can define remote <b>UDF functions</b>. Let's see an example of that."
```

Comment:
Let's add link for BigQuery DataFrames remote functions: https://cloud.google.com/bigquery/docs/samples/bigquery-dataframes-remote-function